Added tesnorizeation for avx2 based gemm.

kimishpatel opened the pull request:
Summary:
Similar to other avx512 tensorization that reduces data:1x4 and kernel:16x4 to output:1x16, this PR introduces similar reduction using avx2 tensorization. It keeps the same API as avx512 so as to not have to introduce a new memory layout for weights.

Test Plan:
on avx2 machine:
python tests/python/contrib/test_gemm_avx2_acc32.py

Reviewers:

Subscribers:

Tasks:

Tags:

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

kimishpatel commented:
Depends on this PR: https://github.com/dmlc/tvm/pull/3981

anijain2305 commented on topi/python/topi/x86/tensor_intrin.py:
```diff
@@ -275,3 +275,97 @@ def _instr(index):
+            vec_b_1 = ins[1].vload([8, 0], "int8x32")
+            vec_one = tvm.const(1, "int16x16")
+            pair_reduction_0 = tvm.call_llvm_intrin('int16x16',
+                                                  'llvm.x86.avx2.pmadd.ub.sw',
```
Align (and other places as well)

anijain2305 commented on topi/python/topi/x86/tensor_intrin.py:
```diff
@@ -275,3 +275,97 @@ def _instr(index):
+                                                  pair_reduction_1, vec_one)
+            if index == 0:
+                ib.emit(outs[0].vstore([0], quad_reduction_0))
+                ib.emit(outs[0].vstore([8], quad_reduction_1))
```
Why have 2 quad_reductions in assembly instead of unrolling in the schedule?

anijain2305 replied:
Oh maybe, you would have then change num_lanes etc

If thats tha case, @yzhliu is working on making it an argument. Maybe we can use that to simplify this.

kimishpatel replied:
@anijain2305. Correct. I wanted to be able to use the same memory layout which can be tensorized for avx2 and avx512.

kimishpatel commented:
@anijain2305, thats a good question. I havent checked the performance of this agains fp32 model we are interested in. I will report the numbers soon. But we do get a lot of speed up on skylake with avx512 and assuming we dont get anything from AVX2 instructions itself due to latency issues you mention, there should still be some perf left from the memory bw persepctive.

anijain2305 commented:
> @anijain2305, thats a good question. I havent checked the performance of this agains fp32 model we are interested in. I will report the numbers soon. But we do get a lot of speed up on skylake with avx512 and assuming we dont get anything from AVX2 instructions itself due to latency issues you mention, there should still be some perf left from the memory bw persepctive.

Ok. This might be useful to do some paper calculations to see if we can get performance theoretically - https://www.agner.org/optimize/instruction_tables.pdf

kimishpatel commented:
@anijain2305, do you mind if I land this? I would like to use it for some of the stuff we are working on.

anijain2305 commented on tests/python/contrib/test_gemm_avx2_acc32.py:
```diff
@@ -0,0 +1,98 @@
+    X = tvm.placeholder((m, k), name='X', dtype="uint8")
+    W = tvm.placeholder((n, k), name='W', dtype="int8")
+
+    #peak = 280 // This needs measurement and description of what this number is for avx2 machine.
```
Please remove the comments

anijain2305 commented:
> @anijain2305, do you mind if I land this? I would like to use it for some of the stuff we are working on.

Yes, please go ahead. The changes LGTM. Would appreciate if later you could update with the performance comparison.

kimishpatel commented:
Thanks @anijain2305, will do so.

kimishpatel commented:
Just as FYI. CI is failing on the test I added as it is dependent on this PR: https://github.com/dmlc/tvm/pull/3981/. I need to land that.

kimishpatel commented:
@anijain2305, would mind merging this please? I dont think I can, plus it seems that my lint changes and addressing your comments might have voided your approval.

anijain2305 commented:
I don't have merge permissions :)
@tqchen can you please take a look?

tqchen commented:
Thanks @kimishpatel @anijain2305
